src/ContactsTabManager.php

```diff
@@ -166,67 +160,78 @@ public function getBlocks(ContactTabInterface $tab, UserInterface $contact = NUL
-          }
-          catch (ContextException $exception) {
-            continue;
-          }
-        }
-
-        // @todo Don't allow block to be added if relationships have not been
-        // satisfied.
         $blocks[$key] = $block;
       }
-
-      $tab->setBlockPlugins($blocks);
     }
 
-    // Verify the tab unless we've been asked not to.
-    if ($verify) {
-      $this->verifyTab($tab, $contact, $blocks);
-
-      // Filter our blocks that don't verify.
-      $blocks = array_filter($blocks, function ($block) {
-        return $block->_contactTabVerified;
-      });
+    if ($contact !== NULL) {
+      $this->applyBlockContext($tab, $blocks, $contact);
     }
 
+    $tab->setBlockPlugins($blocks);
+
     return $blocks;
   }
 
   /**
-   * {@inheritdoc}
+   * Apply block plugin context requirements.
+   *
+   * Where contexts are not able to be matched, blocks will be removed from tab.
+   *
+   * @param \Drupal\contacts\Entity\ContactTabInterface $tab
+   *   The tab entity the blocks are on.
+   * @param \Drupal\Core\Block\BlockPluginInterface[] $blocks
+   *   Blocks to apply the contexts to.
+   * @param \Drupal\user\UserInterface $contact
+   *   The contact we are applying contexts for.
    */
-  public function verifyTab(ContactTabInterface $tab, UserInterface $contact, array $blocks = []) {
-    // Get the block if we don't already have it.
-    if (empty($blocks)) {
-      $blocks = $this->getBlocks($tab, $contact, FALSE);
+  protected function applyBlockContext(ContactTabInterface $tab, array &$blocks, UserInterface $contact) {
+    // Apply context to our block plugins if relevant.
+    foreach ($blocks as $key => &$block) {
+      /* @var \Drupal\Core\Block\BlockPluginInterface $block */
+      if ($block instanceof ContextAwarePluginInterface) {
+        try {
+          // Build our contexts.
+          $contexts = [
+            'user' => new Context(new ContextDefinition('entity:user'), $contact),
+            'subpage' => new Context(new ContextDefinition('string'), $tab->getPath()),
+          ];
+
+          // Gather any relationships.
+          foreach ($tab->getRelationships() as $relationship) {
+            if (isset($contexts[$relationship['source']])) {
+              /* @var \Drupal\ctools\Plugin\RelationshipInterface $plugin */
+              $plugin = $this->relationshipManager->createInstance($relationship['id']);
+              $plugin->setContext('base', $contexts[$relationship['source']]);
+              $contexts[$relationship['name']] = $plugin->getRelationship();
+            }
+          }
 
-      if (empty($blocks)) {
-        return FALSE;
+          // Apply the mappings.
+          $this->contextHandler->applyContextMapping($block, $contexts);
+        }
+        catch (ContextException $exception) {
+          unset($blocks[$key]);
+          continue;
+        }
       }
+
+      // @todo Don't allow block to be added if relationships have not been
+      // satisfied.
```

Comment:
I think this is already done with the code above? In which case this can be removed...